CONTRIBUTING.md

```diff
@@ -10,55 +10,145 @@ For general contribution and community guidelines, please see the [community rep
 
 ## Development
 
+### Building and packaging
+
+To build versions for Linux, macOS and Windows:
+
+```
+./build
+```
+
+Binaries will be placed in `dist/`.
+
+### Running the project from source
+
 Run the project with:
 
 ```
 go run cmd/main.go
 ```
 
-### Testing
+### Unit Testing
 
 Tests are written using [GoConvey](http://goconvey.co/).
 Run tests with `go test -v ./...` or `./test` (for CI).
 
-### Building and packaging
+### Smoke Testing
 
-To build versions for Linux, macOS and Windows:
+Smoke testing is used to verify that the output generated by our `./build` script
+functions as expected.
+
+This is done through running the generated binaries against production or custom
+providers and verifying the output manually.  
+
+### Generating Mock Providers
+
+For testing purposes, it is often necessary to create a "mock" provider which can verify 
+input and output. This can be done by creating a script which can be run by Summon and
+placing it in the default path, `/usr/local/lib/summon/`
 
+For example, we can create a provider that simply returns the literal variable that is
+passed into it:
+
+1. Create a shell script in your Summon path called `mock`:
+```bash
+sudo <nano, vim, etc.> /usr/local/lib/summon/mock
 ```
-./build
+
+2. Add the following to your script:
+```bash
+#!/bin/bash -e
+echo "$@"
 ```
 
-Binaries will be placed in `dist/`.
+3. Make your new script executable
+```bash
+sudo chmod +x /usr/local/lib/summon/mock
+```
 
-### Releasing
+4. Use your 'provider' with the following command
+```bash
+summon -p mock --yaml 'FOO: !var bar' /bin/bash -ec 'echo $FOO'
+```
+
+This mock provider can then be used against locally generated binaries or the source
+itself, as needed.
+
+## Releasing
 
 The following checklist should be followed when creating a release:
 
 - [ ] Open a PR with the following changes for `master` branch and wait for it to be merged:
   - [ ] Bump release version in [`pkg/summon/version.go`](pkg/summon/version.go).
   - [ ] Bump version in [`CHANGELOG.md`](CHANGELOG.md).
+  
+- [ ] [`Build`](./build) the release. Output it placed in the `dist` folder.
+   - [ ] Run smoke tests using the binaries in their respective output directories on
+         their appropriate platform.
+        - Linux:   `dist/summon-linux_linux_amd64`
+        - MacOS:   `dist/summon_darwin_amd64`
+        - Windows: `dist/summon_windows_amd64`
```

Comment:
This is also done by jenkins - no need to build locally but it's fine to leave it here as alternatives